### DISCUSSION TITLE
apollo async atom

### DISCUSSION THREAD

**Discussion by mzalevski** [pmndrs/jotai#1069] [Q&A] [answered]
Hi, I'd love to get some advice on the following:

I have a global atom for the logged in user:
```
export const currentUserAtom = atom<MeQuery["me"]>(null);
```
the data is fetched after login:
```
const [, setCurrentUser] = useAtom(currentUserAtom);
const { data, error } = useMeQuery();

useEffect(() => {
  setCurrentUser(data.me);
}, [data]);
```
I'd like to incorporate the fetching logic into the `currentUserAtom` using async atom (https://jotai.org/docs/basics/async).

I'm not sure how to get this right. The docs are not sufficient for me.

I tried something like this:
```
export const currentUserAtom = atom(async () => {
  const client = useApolloClient();
  const { data } = await client.query<MeQuery>({ query: MeDocument });
  return data.me;
});
```
but it loads forever (suspense fallback is not showing) and I'm not even using the `get` function, because there is no other atom I can pass to it. I tried separating a promise atom to then use the `get` function on it, but it didn't work and felt wrong.

**Comment by dai-shi** [comment]
Hi! Thanks for using the lib.
Atom logic is outside React, so you can't use hooks in `get` function.
(Though, I wonder how it causes an infinite loop...)

Hope @Aslemammad can help on this specific problem with Apollo as he developed [jotai-apollo](https://github.com/Aslemammad/jotai-apollo).

**Reply by Aslemammad** [reply]
@mzalevski Could you share a reproduction so I can help you with that?

**Reply by mzalevski** [reply]
@Aslemammad sure, I will share the repo in a moment. thanks :)

**Reply by mzalevski** [reply]
@Aslemammad https://github.com/mzalevski/jotai-apollo

**Reply by Aslemammad** [reply]
Here's a demonstration, please try cleaning it, I just wanted to give you an idea, if you give an atom a promise, jotai will suspend and handle it for you, so you should wrap your Components with Suspense!

```tsx
// hooks.ts
import { ApolloQueryResult, gql } from "@apollo/client";
import { useAtomCallback } from "jotai/utils";
import { useEffect, useState } from "react";
import { client } from "./App";
import { currentCountryAtom } from "./atoms";
import { Country } from "./types";

export const useGetCurrentCountry = () => {
  const [promise, setPromise] = useState<any>(null);

  const setResult = useAtomCallback((_, set) => {
    set(currentCountryAtom, promise);
  });

  const query = gql`
    query {
      country(code: "PL") {
        code
        name
      }
    }
  `;
  
  // changing state using useEffect is not ideal, keep that in mind!
  useEffect(() => {
    setPromise(
      client
        .query({ query })
        .then(
          (resp: ApolloQueryResult<{ country: Country }>) => resp.data?.country
        )
    );
  }, [client]);

  useEffect(() => {
    setResult();
  }, [promise]);
};
```
```tsx
// App.tsx
import { BrowserRouter, Route, Routes } from "react-router-dom";
import { ApolloClient, InMemoryCache, ApolloProvider } from "@apollo/client";
import { HomePage } from "./components/home-page";
import { OtherPage } from "./components/other-page";
import { Wrapper } from "./components/wrapper";
import { Suspense } from "react";

export const client = new ApolloClient({
  uri: "https://countries.trevorblades.com",
  cache: new InMemoryCache(),
});

export const App = () => (
  <ApolloProvider client={client}>
    <BrowserRouter>
      <Suspense fallback={<div>Loading...</div>}>
        <Wrapper>
          <Routes>
            <Route index element={<HomePage />} />
            <Route path="other" element={<OtherPage />} />
          </Routes>
        </Wrapper>
      </Suspense>
    </BrowserRouter>
  </ApolloProvider>
);

```

**Reply by Aslemammad** [reply]
if you want to do it without hooks, see [this](https://github.com/Aslemammad/jotai-apollo/blob/c13d0f7349736e0bb0051cb8f931dffb7118c60a/src/atomWithQuery.ts#L81). 
There, the atom does not suspend, because I don't give it a promise, but if you want it to suspend, create similar atom, and give the client query promise to an atom that you want to use!

**Comment by mzalevski** [comment, marked as answer]
ok, thanks to 2 insights:
1. Atom logic is outside React, so you can't use hooks in get function.
2. if you give an atom a promise, jotai will suspend and handle it for you

I figured it out. see: https://github.com/mzalevski/jotai-apollo

```ts
export const currentCountryAtom = atom<Promise<Country>>((get) => {
  return get(clientAtom)
    .query<{ country: Country }>({ query: COUNTRY })
    .then((result) => result.data.country);
});
```

thanks @dai-shi @Aslemammad
